File: app/MMarket/c/src/pgs.c
// Copyright 13-Dic-2021 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "pgs.h"
#include <stdlib.h>
#include "dmc/err.h"
#include "dmc/str.h"
#include "dmc/sys.h"
#include "dmc/cgi.h"
#include "dmc/cryp.h"
#include "dmc/js.h"
#include "data/cts.h"
#include "pgs/main.h"
#include "pgs/descriptionPg.h"
#include "pgs/resultsPg.h"
#include "pgs/hotPg.h"
#include "pgs/cosPg.h"
#include "pgs/historicPg.h"
#include "pgs/operationsPg.h"
#include "pgs/changePass.h"

static char *server_source_process (Mchar *mrq) {
  char *source = cgi_rq_string(mrq, "source");
  if (str_eq(source, "Main")) return main_process(mrq);
  else if (str_eq(source, "Description")) return descriptionPg_process(mrq);
  else if (str_eq(source, "ResultsPg")) return resultsPg_process(mrq);
  else if (str_eq(source, "HotPg")) return hotPg_process(mrq);
  else if (str_eq(source, "CosPg")) return cosPg_process(mrq);
  else if (str_eq(source, "HistoricPg")) return historicPg_process(mrq);
  else if (str_eq(source, "OperationsPg")) return operationsPg_process(mrq);
  else if (str_eq(source, "ChangePass")) return changePass_process(mrq);
  else return FAIL(str_f("Unexpected value for 'source': %s", source));
}

char *pgs_process(char *rq) {
  int ix = str_cindex(rq, ':');

  //............................................................. CONNECTION
  if (ix == -1) {
    return(cgi_connect(rq));
  }

  //......................................................... AUTHENTICATION
  if (ix == 0) {
    char *key = cryp_key(cts_app_name(), cgi_klen());
    char *data = cryp_decryp(key, rq + 1);
    Achar *parts = str_csplit(data, ':');
    return(cgi_authentication(
      key,
      achar_get(parts, 0),
      achar_get(parts, 1),
      *(achar_get(parts, 2)) == '1'
    ));
  }

  //............................................................ NORMAL DATA
  Achar *parts = str_csplit(rq, ':');
  char *session_id = achar_get(parts, 0);
  char *data = achar_get(parts, 1);
  char *con_key = "";
  if (achar_size(parts) > 2) {
    con_key = data;
    data = achar_get(parts, 2);
  }

  char *key = ochar_nsome(cgi_set_key(session_id, con_key));
  if (key) {
    return server_source_process(js_ro(cryp_decryp(key, data)));
  }

  return(cgi_rp_expired());
}

File: app/KtMMarket/go/pgs/pgs.go
// Copyright 01-Jul-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Pages hub.
package pgs

import (
	"github.com/dedeme/KtMMarket/cts"
	"github.com/dedeme/KtMMarket/pgs/changePassPg"
	"github.com/dedeme/KtMMarket/pgs/cosPg"
	"github.com/dedeme/KtMMarket/pgs/descriptionPg"
	"github.com/dedeme/KtMMarket/pgs/historicPg"
	"github.com/dedeme/KtMMarket/pgs/hotPg"
	"github.com/dedeme/KtMMarket/pgs/mainPg"
	"github.com/dedeme/KtMMarket/pgs/operationsPg"
	"github.com/dedeme/KtMMarket/pgs/resultsPg"
	"github.com/dedeme/KtMMarket/pgs/simulationsPg"
	"github.com/dedeme/ktlib/cgi"
	"github.com/dedeme/ktlib/cryp"
	"github.com/dedeme/ktlib/js"
	"github.com/dedeme/ktlib/str"
	"github.com/dedeme/ktlib/sys"
)

func Process(rq string) (rp string) {
	defer func() {
		if err := recover(); err != nil {
			rp = sys.Fail(str.Fmt("%v", err))
			ix1 := str.Index(rp, ":\n")
			if ix1 != -1 {
				ix2 := str.IndexFrom(rp, "runtime/panic.go", ix1)
				if ix2 != -1 {
					ix2 = str.IndexFrom(rp, "\n", ix2)
				}
				if ix2 != -1 {
					rp = rp[:ix1+2] + rp[ix2+1:]
				}
			}
		}
	}()

	ix := str.Index(rq, ":")

	// CONNECTION --------------------------------------------------------------

	if ix == -1 {
		return cgi.Connect(rq)
	}

	// AUTHENTICATION ----------------------------------------------------------

	if ix == 0 {
		key := cryp.Key(cts.AppName, cgi.Klen)
		data := cryp.Decode(key, rq[1:])
		ps := str.Split(data, ":")
		return cgi.Authentication(key, ps[0], ps[1], ps[2] == "1")
	}

	// NORMAL DATA -------------------------------------------------------------

	sessionId := rq[:ix]
	var conKey string
	rest := rq[ix+1:]
	ix = str.Index(rest, ":")
	if ix != -1 {
		conKey = rest[:ix]
		rest = rest[ix+1:]
	}
	comKey, ok := cgi.GetComKey(sessionId, conKey)
	if ok {
		j := cryp.Decode(comKey, rest)
		data := js.Ro(j)
		return processHub(comKey, data)
	} else {
		return cgi.RpExpired()
	}
}

func processHub(ck string, rq cgi.T) string {
	source := js.Rs(rq["source"])
	switch source {
	case "Main":
		return mainPg.Process(ck, rq)
	case "Description":
		return descriptionPg.Process(ck, rq)
	case "ResultsPg":
		return resultsPg.Process(ck, rq)
	case "SimulationsPg":
		return simulationsPg.Process(ck, rq)
	case "HotPg":
		return hotPg.Process(ck, rq)
	case "CosPg":
		return cosPg.Process(ck, rq)
	case "HistoricPg":
		return historicPg.Process(ck, rq)
	case "OperationsPg":
		return operationsPg.Process(ck, rq)
	case "ChangePass":
		return changePassPg.Process(ck, rq)
	default:
		panic("Value of source (" + source + ") is not valid")
	}
}
